### PR TITLE
Fix user account deletion not working

### LINKED ISSUE #26210: Delete user account functionality broken.
### Description

Steps to repro.
1. Login to gitea
2. Go to settings.
3. Go to Accounts.
4. Go to `Delete Your Account`.
5. Now click on 
image without entring password. 
6. We get 500 error.

### Gitea Version

1.21.0+dev-412-gce27de4d4

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

image
image
image

### Git Version

1.21.0+dev-412-gce27de4d4

### Operating System

_No response_

### How are you running Gitea?

locally and try.gitea.io

### Database

None

### PR DESCRIPTION
**Pull request by puni9869**
as title:
Fixes:  #26210

### CODE REVIEW

**Comment by puni9869** [comment]
@techknowlogick  Need your feedback on this approach.

**Comment by puni9869** [comment]
I think we should limit the use of `ctx.ServerError("<namespace>", err)`

**Comment by lunny** [comment]
No, I don't think it's right. We need to enumerate all possible errors.

**Comment by puni9869** [comment]
> No, I don't think it's right. We need to enumerate all possible errors.

But it is limited to password check and do you want the whole `user_model` should be checked before proceeding to delete the account

**Comment by lunny** [comment]
> > No, I don't think it's right. We need to enumerate all possible errors.
> 
> But it is limited to password check and do you want the whole `user_model` should be checked before proceeding to delete the account.
> 
> Could you suggest something in one or two liner the approach. It will be easy for me to understand.

At least, to resolve #26210, you could catch paswrod is invalid error and display it in the UI and let other errors return 500.

Some errors are affected by wrong user's input, some are from system problems. We need to diff these two kinds of errors.

**Comment by puni9869** [comment]
> ed to diff these two kinds of errors.

Ohh, I got it now. Sending the fix in a few

**Comment by puni9869** [comment]
image

**Comment by lunny** [comment]
see https://github.com/go-gitea/gitea/issues/26210#issuecomment-1657193397

**Comment by puni9869** [comment]
> see [#26210 (comment)](https://github.com/go-gitea/gitea/issues/26210#issuecomment-1657193397)

I will share  couple samples of implementation in the followups. Apologies for this,

**Comment by yp05327** [comment]
Is it possible to reuse `handleSignInError`?
And this is strange:
image

**Comment by puni9869** [comment]
> Is it possible to reuse `handleSignInError`? And this is strange: image

Yes, I have used it. If you see the commits but the error scope is large. I am getting panic in below snipped.

```go
_, _, err := auth.UserSignIn(ctx.Doer.Name, ctx.FormString("password"))
if err != nil {
    if user_model.IsErrUserNotExist(err) {
	    loadAccountData(ctx)
	    ctx.RenderWithErr(ctx.Tr("form.enterred_invalid_password"), tplSettingsAccount, nil)
    } else if errors.Is(err, err.(db.ErrUserPasswordNotSet)) {
	    ctx.RenderWithErr(ctx.Tr("form.password_not_set_on_user_account"), tplSettingsAccount, nil)
    } else if errors.Is(err, err.(db.ErrUserPasswordInvalid)) {
	    ctx.RenderWithErr(ctx.Tr("form.enterred_invalid_password"), tplSettingsAccount, nil)
    } else if user_model.IsErrUserProhibitLogin(err) {
	    ctx.RenderWithErr(ctx.Tr("auth.prohibit_login"), tplSettingsAccount, nil)
    } else {
	    ctx.ServerError("UserSignIn", err)
    }
    return
}
```

File: routers/web/user/setting/account.go [hunk 1]
```diff
@@ -234,7 +234,11 @@ func DeleteEmail(ctx *context.Context) {
 func DeleteAccount(ctx *context.Context) {
 	ctx.Data["Title"] = ctx.Tr("settings")
 	ctx.Data["PageIsSettingsAccount"] = true
-
+	passwd := ctx.FormString("password")
+	if len(passwd) == 0 || !ctx.Doer.ValidatePassword(passwd) {
```
Comment (techknowlogick) [inline routers/web/user/setting/account.go]: Thanks for making this change. `auth.UserSignIn` below should return `ErrUserPasswordInvalid` if the password is invalid, so perhaps that could be checked in an if conditional instead (pseudocode `err == ErrUserPasswordInvalid`)?

Comment (puni9869) [reply routers/web/user/setting/account.go]: I will add this check too. But lil bit confused with the error stack that I have create in above comments are right way of doing it? Need a bit of help there. 
The committed solution will just works as fine. But there are other errors thats may occurs that needs to be covered.

**Review by delvh — APPROVED** [review]

File: routers/web/user/setting/account.go [hunk 2]
```diff
@@ -234,7 +234,11 @@ func DeleteEmail(ctx *context.Context) {
...
+		ctx.RenderWithErr(ctx.Tr("form.enterred_invalid_password"), tplSettingsAccount, nil)
+		return
+	}
```
Comment (wxiaoguang) [inline routers/web/user/setting/account.go]: How does it work with external auth sources? eg: LDAP, the `passwd` field is empty.

(the first question is: should Gitea support deleting LDAP accounts?)

Comment (puni9869) [reply routers/web/user/setting/account.go]: @lunny Need your help on this.

Comment (lunny) [reply routers/web/user/setting/account.go]: Maybe we need to check whether password is set before check popup the password dialog.

Comment (puni9869) [reply routers/web/user/setting/account.go]: You mean we need an api call from frontend that will check the password is set on the account?

Comment (puni9869) [reply routers/web/user/setting/account.go]: Thinking that,  there are still some edge cases remaining before we implement the solution around this 500 error.
I'm keeping this PR open for the further discussion and will raise a complete new PR.

**Pull request closed without merging**